### PR TITLE
Cis benchmark 1.22

### PR DESCRIPTION
**Pull request by ndowmon**
This stuff was a bit of a head scratcher. Thanks @mknoedel for your help yesterday.

Let me know if you'd like more explanation anywhere.

### CODE REVIEW

File: src/steps/active-directory/client.ts [hunk 1]
```diff
@@ -24,7 +25,55 @@ export interface GroupMember extends DirectoryObject {
...
+            description: `Unable to fetch data from ${path}. See https://github.com/JupiterOne/graph-azure/blob/master/docs/jupiterone.md#permissions for more information about optional permissions for this integration.`,
+          });
+          return;
+        }
+      }
+
+      // This endpoint is brittle, since it behaves differently whether the default directory (tenant) is a "personal"
+      // account or a "school/work" account. In order to protect us from execution failures during an important active directory
+      // step, we'll never throw an error here but _explicitly_ warn operators (developers) using logger.error, and
+      // also send a message to sentry via logger.onError.
```
Comment (ndowmon) [inline src/steps/active-directory/client.ts:64]:
```suggestion
      // also send a message to sentry via logger.onFailure.
```

File: src/steps/active-directory/__recordings__/active-directory-steps_52759203/recording.har [hunk 2]
```diff
@@ -237,7 +237,7 @@
...
...ryObjects/992d7bbe-b367-459c-a10f-cf3fd16103ab/Microsoft.DirectoryServices.Organization\",\"id\":\"992d7bbe-b367-459c-a10f-cf3fd16103ab\",\"deletedDateTime\":null,\"businessPhones\":[],\"city\":null,\"country\":null,\"countryLetterCode\":\"US\",\"createdDateTime\":\"2020-07-16T17:54:20Z\",\"displayName\":\"Default Directory\",\"isMultipleDataLocationsForServicesEnabled\":null,\"marketingNotificationEmails\":[],\"onPremisesLastSyncDateTime\":null,\"onPremisesSyncEnabled\":null,\"postalCode\":null,\"preferredLanguage\":\"en\",\"securityComplianceNotificationMails\":[],\"securityComplianceNotificationPhones\":[],\"state\":null,\"street\":null,\"technicalNotificationMails\":[\"ndowmon@gmail.com\"],\"tenantType\":\"AAD\",\"directorySizeQuota\":{\"used\":190,\"total\":50000},\"privacyProfile\":null,\"assignedPlans\":[{\"assignedDateTime\":\"2020-09-15T15:38:22Z\",\"capabilityStatus\":\"Enabled\",\"service\":\"WindowsAzure\",\"servicePlanId\":\"fca3e605-0754-4279-8504-3f1229f29614\"}],\"provisionedPlans\":[],\"verifiedDomains\":[{\"capabilities\":\"Email, OfficeCommunicationsOnline\",\"isDefault\":true,\"isInitial\":true,\"name\":\"ndowmongmail.onmicrosoft.com\",\"type\":\"Managed\"}]}]}"
```
Comment (mknoedel) [inline src/steps/active-directory/__recordings__/active-directory-steps_52759203/recording.har:240]: Your email is in here if you care

File: src/steps/active-directory/client.ts [hunk 3]
```diff
@@ -24,7 +25,55 @@ export interface GroupMember extends DirectoryObject {
...
+      //
+      // In the future when this endpoint is better understood, we can improve the handling here.
+      this.logger.error(err);
+      try {
+        (this.logger as any)?.onFailure({ err });
```
Comment (mknoedel) [inline src/steps/active-directory/client.ts:69]: Does this throw a message to the jobLog or something?

File: src/steps/resource-manager/network/index.test.ts [hunk 4]
```diff
@@ -114,9 +114,6 @@ describe('network steps', () => {
     recording = setupAzureRecording({
       directory: __dirname,
       name: 'network-steps',
-      options: {
-        recordFailedRequests: true,
-      },
```
Comment (ndowmon) [inline src/steps/resource-manager/network/index.test.ts:119]: Might seem random for me to have removed this, but this test suite actually calls `fetchAccount()` and hence calls the new endpoint (unsuccessfully, because the recorded token == `[REDACTED]`).

It was causing some failed API calls to be recorded and would have caused an unnecessary change in `__recordings`.

File: src/steps/active-directory/client.ts [hunk 5]
```diff
@@ -24,7 +25,55 @@ export interface GroupMember extends DirectoryObject {
...
+      );
+    } catch (err) {
+      if (err instanceof IntegrationProviderAPIError) {
+        if (
+          err._cause?.message ===
+          'You cannot perform the requested operation, required scopes are missing in the token.'
+        ) {
+          this.logger.publishEvent({
+            name: 'missing_optional_permission',
+            description: `Unable to fetch data from ${path}. See https://github.com/JupiterOne/graph-azure/blob/master/docs/jupiterone.md#permissions for more information about optional permissions for this integration.`,
```
Comment (mknoedel) [inline src/steps/active-directory/client.ts:55]: 💯

File: src/steps/active-directory/__recordings__/active-directory-steps_52759203/recording.har (hunk 2 above)
Comment (ndowmon) [reply src/steps/active-directory/__recordings__/active-directory-steps_52759203/recording.har:240]: I'm OK with it. It's everywhere...

**Review by mknoedel — DISMISSED** [review]
Looks great

File: src/steps/active-directory/client.ts (hunk 3 above)
Comment (ndowmon) [reply src/steps/active-directory/client.ts:69]: This is a callback we [instrument in the managed project](https://bitbucket.org/jupiterone/jupiter-managed-integration/src/ffbf0dc34cadfab4de3f1ec6156e6245bd981760/packages/runtime/src/executeIntegration.ts#lines-75). It's used to send errors to Sentry, or whatever error handling system we'll use in the future.

This is a hacky way to get at it but I'm hoping that unhandled failures here will help us understand how to actually handle exceptions, and throwing in sentry seems like a good way to get our eyes on this.

File: src/steps/active-directory/client.ts [hunk 6]
```diff
@@ -24,7 +25,55 @@ export interface GroupMember extends DirectoryObject {
...
+        this.client.api(path),
+      );
+    } catch (err) {
+      if (err instanceof IntegrationProviderAPIError) {
+        if (
+          err._cause?.message ===
+          'You cannot perform the requested operation, required scopes are missing in the token.'
+        ) {
+          this.logger.publishEvent({
+            name: 'missing_optional_permission',
```
Comment (aiwilliams) [inline src/steps/active-directory/client.ts:54]: Is it missing if it's optional?

File: src/steps/active-directory/client.ts [hunk 7]
```diff
@@ -24,7 +25,55 @@ export interface GroupMember extends DirectoryObject {
...
+      // step, we'll never throw an error here but _explicitly_ warn operators (developers) using logger.error, and
+      // also send a message to sentry via logger.onError.
+      //
+      // In the future when this endpoint is better understood, we can improve the handling here.
+      this.logger.error(err);
+      try {
+        (this.logger as any)?.onFailure({ err });
+      } catch (err) {
+        // pass
+      }
```
[the remark this answers is not in the document]
Comment (ndowmon) [reply src/steps/active-directory/client.ts:72]: Good question. Should've been `(this.logger as any).onFailure?({ err });` but since it's try-catch anyway there's really no need for optional chaining.

File: src/steps/active-directory/client.ts (hunk 6 above)
Comment (ndowmon) [reply src/steps/active-directory/client.ts:54]: I kind of think yes. I was originally planing to use `missing_permission` but opted for this instead.

Comment (aiwilliams) [reply src/steps/active-directory/client.ts:54]: I'm still a fan of just `auth`, but we'll have to review the world anyway when we work to make things consistent.

**Review by aiwilliams — APPROVED** [review]

File: src/steps/active-directory/client.ts (hunk 6 above)
Comment (ndowmon) [reply src/steps/active-directory/client.ts:54]: https://github.com/JupiterOne/graph-azure/issues/230
